### PR TITLE
[CA-107] Doc structural update + Added a test

### PR DESCRIPTION
**Pull request by tejas-hosamani**
## Structural changes:
- Converted JSON files to YAML for better readability.
- However, `APIs.json` is as is because `swagger-ui-watcher` package doesn't support YAML file creation as of now.

## Tests update:
- Added an additional test to make sure APIs are queriable from Doc page. In other words, making sure 'Try it out' feature works and returns API response. 

Please review whether the code is clean enough to merge. Thanks!

### CODE REVIEW

**Comment by CodingCreate101** [comment]
Oops. New test failed.

File: src/api/docs/doc.test.js [hunk 1]
```diff
@@ -1,10 +1,10 @@
 const puppeteer = require('puppeteer');
-const { PORT } = require('../../config');
+const { HOST, PORT } = require('../../config');
 const app = require('../../app').listen(PORT);
 
 let browser;
 let page;
-const BASE_URL = `http://localhost:${PORT}`;
+const BASE_URL = `http://${HOST}:${PORT}`;
```
Comment (ewenjo) [inline src/api/docs/doc.test.js]: What's the reason for this change? All tests are run locally, localhost is fine.

File: src/api/docs/doc.test.js [hunk 2]
```diff
@@ -17,15 +17,41 @@ afterAll(async () => {
...
+      await page.waitForSelector('div > span > #operations-Contributors-get_contributors > .opblock-summary > .opblock-summary-description');
+      await page.click('div > span > #operations-Contributors-get_contributors > .opblock-summary > .opblock-summary-description');
+
+      await page.waitForSelector('.opblock-body > .opblock-section > .opblock-section-header > .try-out > .btn');
+      await page.click('.opblock-body > .opblock-section > .opblock-section-header > .try-out > .btn');
+
+      await page.waitForSelector('#operations-Contributors-get_contributors > div > .opblock-body > .execute-wrapper > .btn');
+      await page.click('#operations-Contributors-get_contributors > div > .opblock-body > .execute-wrapper > .btn');
+
+      await page.waitForSelector('.response > .col > div > .highlight-code > .microlight');
```
Comment (ewenjo) [inline src/api/docs/doc.test.js]: This won't actually check if the query is successful.

File: src/api/docs/doc.test.js [hunk 3]
```diff
@@ -17,15 +17,41 @@ afterAll(async () => {
...
+    } catch (e) {
+      // eslint-disable-next-line no-console
+      console.log(e);
+      // console.log('TEST FAILED: Could not query /contributors API from Doc page.');
```
Comment (ewenjo) [inline src/api/docs/doc.test.js]: Include the error in the console.log

File: src/api/docs/doc.test.js (hunk 1 above)
Comment (tejas-hosamani) [reply src/api/docs/doc.test.js]: You are right. It wasn't necessary.

In fact, it is necessary that the host is `localhost` because of the 3rd test. I was thinking about this in the opposite direction.

File: src/api/docs/doc.test.js (hunk 3 above)
Comment (tejas-hosamani) [reply src/api/docs/doc.test.js]: That error has a lot of content. I kept the console minimum and wrote a custom error message. 
I could use` e.message` but that doesn't say anything about the actual error that we are looking for. What do you think?

File: src/api/docs/doc.test.js (hunk 2 above)
Comment (tejas-hosamani) [reply src/api/docs/doc.test.js]: Yes, not directly. But this element only exists after a successful request. By that, we can conclude the request was successful.

Do you have any suggestion to directly check the request's success? I 'll look into this more in the meantime.

Comment (ewenjo) [reply src/api/docs/doc.test.js]: Sorry, I meant "successful request" in my original comment. 

Might not be necessary to check it because we already have tests making sure the routes works, but we can go the extra mile and check for status code in the response body. What do you think?

File: src/api/docs/doc.test.js (hunk 3 above)
Comment (ewenjo) [reply src/api/docs/doc.test.js]: We should log the whole error. When an error occurs, we want as much information as possible to debug and solve the problem.

File: src/api/docs/doc.test.js (hunk 2 above)
Comment (tejas-hosamani) [reply src/api/docs/doc.test.js]: This check is not about whether APIs work or not. It's whether the API doc can call them or not through UI. Because, that's what the user is going to use(front-end devs). That's why I thought it's necessary. It's the part where end-to-end testing comes in, I suppose(?)

The second part actually makes a lot of sense. It is an accurate thing to do. That's what I originally considered doing. But I am at the limit of my knowledge about Puppeteer. I've been looking through their documentation. I'll update it with actual response code check. That's a better thing to do.

File: src/api/docs/doc.test.js (hunk 3 above)
Comment (tejas-hosamani) [reply src/api/docs/doc.test.js]: That makes sense. Alright, I will log the complete error.

File: src/api/docs/doc.test.js (hunk 2 above)
Comment (ewenjo) [reply src/api/docs/doc.test.js]: Sounds good, we should make sure the CI tests pass before adding anything else. Do you have any ideas why the tests are failing?

Comment (tejas-hosamani) [reply src/api/docs/doc.test.js]: Probably because of the `const BASE_URL = 'http://${HOST}:${PORT}';`
Depends on CI's env HOST variable. If it's anything other than `localhost`, that's definitely the reason. And changing my code back to const BASE_URL = `http://localhost:${PORT}`; will fix that.

If that's not the case, I am not sure. We will get more info once I push with above-mentioned change. 
Let me actually do that right now.

Comment (ewenjo) [reply src/api/docs/doc.test.js]: I was thinking that the localtest URL is hard-coded on port 3000, and the travis ci test can possibly run on other ports. Try changing the host back to "localhost" and see if that fixes anything.

Comment (tejas-hosamani) [reply src/api/docs/doc.test.js]: Yup, that did it. :)

**Review by ewenjo — DISMISSED** [review]
Looks good so far!

**Comment by tejas-hosamani** [comment]
Oh boy, this update took a lot of time!

**Comment by spiray** [comment]
@CodingCreate101 I was thinking about this over the last few days? Does swagger not support versioning? The versions selection seems like an unnecessary extra step to me. Any ideas?

**Comment by ewenjo** [comment]
The last test need some refactoring, this is my take on it.

```javascript
const click = async (selector, p) => {
  const element = await p.waitForSelector(selector);
  await element.click();
};

describe('GET API response through API Doc', () => {
  it('Should be able to query /contributors API from Doc page.', async () => {
    await page.goto(BASE_URL);

    await click('body > ul > li > a', page);
    await click('#operations-Contributors-get_contributors', page);
    await click('.opblock-body > .opblock-section > .opblock-section-header > .try-out > .btn', page);
    await click('.opblock-body > .execute-wrapper > .btn', page);

    await page.waitForSelector('.live-responses-table tbody .response-col_status');
    const statusCodeText = await page.$eval('.live-responses-table tbody .response-col_status', el => el.textContent);
    const statusCode = parseInt(statusCodeText, 10);

    expect(statusCode).toBe(200);
  });
});
```

Also, you should change BASE_URL to:
```javascript
const BASE_URL = `http://localhost:${PORT}/apiDocs`;
```

**Comment by tejas-hosamani** [comment]
> Does swagger not support versioning? The versions selection seems like an unnecessary extra step to me. Any ideas?

I did actually look around for that when I first thought about version maintenance. It is probably possible but we'd have to get a little deep into Swagger React APIs. It's not available at YAML files-level. 
Also, one more question I should have asked is, whether we need to maintain the API versions. I should have asked before making that structure. It's better now than later. 

It is an extra step.. surely. Here's my suggestion:
- The root route, when visited, for example, `localhost/`, redirects the user to default API Doc(latest-stable version APIs Doc.) that we set, ex: `localhost/v1.0.0`. Or it could just be default doc itself, instead of redirecting them.
- `localhost/versions` be the route where we display versions list.
- We could then add this link within each Doc Info section for users to explore other versions.

OR - we could just redirect the user from root router(currently it's displaying `{"msg":"Check out /contributors, /faq, /history and /video"}`) to default API version(Like `localhost/apiDocs/v1.0.0`). And let everything be the way it is.

What do you guys think?

**Comment by tejas-hosamani** [comment]
Refactored the last test.

File: src/api/docs/doc.test.js [hunk 4]
```diff
@@ -18,58 +18,36 @@ afterAll(async () => {
 
 describe('GET /apiDocs/', () => {
   it('Should display "Available versions" text on page.', async () => {
-    await page.goto(`${BASE_URL}/apiDocs/`);
+    await page.goto(`${BASE_URL}/`);
```
Comment (spiray) [inline src/api/docs/doc.test.js]: The template literal is unnecessary here.

**Review by spiray — CHANGES_REQUESTED** [review]

**Comment by spiray** [comment]
@CodingCreate101 I like the idea of having the docs at the root. I don't think we should make ourselves crazy about versioning at this point as this will likely be a very basic API and wont change much overtime. If we are doing separate versions I would make it `/docs/<version>` and `/docs/versions`

File: src/api/docs/doc.test.js (hunk 4 above)
Comment (tejas-hosamani) [reply src/api/docs/doc.test.js]: Sorry, I did not understand what you meant. Could you please explain a bit?

Comment (spiray) [reply src/api/docs/doc.test.js]: That line can be 
```js
await page.goto(BASE_URL);
```
and you can add the slash to the BASE_URL var.

Comment (ewenjo) [reply src/api/docs/doc.test.js]: Change the line to:
```javascript 
await page.goto(BASE_URL);
`````
There is no need for the extra slash, so you don't have to use tempalte literals

**Comment by tejas-hosamani** [comment]
I was thinking in similar lines. So, leave it as is for now then? The structure.

File: src/api/docs/doc.test.js (hunk 4 above)
Comment (tejas-hosamani) [reply src/api/docs/doc.test.js]: Ah. That's right. Thanks!

**Comment by spiray** [comment]
@ewenjo What do you think about having the docs at the root?

**Comment by ewenjo** [comment]
I think the root should be reserved for API routes and docs should have a `/docs` prefix.
I like the look of `/docs/versions` and `/docs/<version>`. This would also allow us to add new versions in the future without changing too much.

Would also be nice if `/docs` redirected to the lastest version.

If it's not obvious, I think we should also rename `/apiDocs` to just `/docs`.

**Comment by tejas-hosamani** [comment]
Could you please elaborate on "root should be reserved for API routes"?
Currently, the domain root is displaying:
```
{"msg":"Check out /contributors, /faq, /history and /video"}
```
It's kind of tiny documentation on its own.

**Comment by ewenjo** [comment]
I'm  just saying that we shouldn't have docs at root. eg. `localhost/v1.0.0`.

The root shouldn't be empty, nor should it just redirect somewhere. Currently it shows some endpoints you can check out, which is fine at the moment. In the future when the API is more complex and have many routes, we will have to change it.

**Comment by tejas-hosamani** [comment]
Alright. So, to clarify..
We're gonna rename `apiDocs` to `docs`. And removing template literal. 

But we are left with that extra step that we were discussing. So, I 'll redirect `/docs` to latest version of Docs. And move version list to `/docs/versions`.
So, we have `/docs/versions` and `/docs/<version>` after all. :)

Is that cool? Did I miss anything?

 @ewenjo @spiray

**Comment by ewenjo** [comment]
Yes, that looks good. Btw, can we use `v1`, `v2` `v3`... instead of `v1.0.0`?
We're not going to change the API that often

**Comment by tejas-hosamani** [comment]
Alright, sounds good.

**Review by spiray — DISMISSED** [review]
Looks good!

**Comment by tejas-hosamani** [comment]
`CodingCreate101 dismissed ewenjo’s stale review via 933ca49`
What does it mean?

**Comment by spiray** [comment]
That just means that when you add a commit a dismisses his review done prior. It happens automatically and isn't something you did.

File: Swag/Info.yaml [hunk 5]
```diff
@@ -0,0 +1,12 @@
+Info:
+  version: 1.0.0
```
Comment (ewenjo) [inline Swag/Info.yaml]: Should this be v1?

File: docs/APIs.json [hunk 6]
```diff
@@ -3,7 +3,7 @@
   "info": {
     "version": "1.0.0",
```
Comment (ewenjo) [inline docs/APIs.json]: Should this also be v1?

File: Swag/Info.yaml (hunk 5 above)
Comment (tejas-hosamani) [reply Swag/Info.yaml]: Yes, I just checked. It can be "user-defined". I thought it isn't. :)
Good catch.. hahha.

Comment (tejas-hosamani) [reply Swag/Info.yaml]: Done!

**Review by ewenjo — APPROVED** [review]

**Review by spiray — APPROVED** [review]
